File: sdk/python/kraken/voice.py
"""Sync voice operations — transcription & text-to-speech."""

from __future__ import annotations

from typing import TYPE_CHECKING, Literal

from kraken.models import TranscriptionResult

if TYPE_CHECKING:
    from kraken._transport import Transport

VoiceId = Literal[
    "alloy", "ash", "ballad", "coral", "echo", "fable", "nova", "onyx", "sage", "shimmer"
]
AudioFormat = Literal["mp3", "opus", "aac", "flac", "wav", "pcm"]


class Voice:
    """Voice API — transcribe audio and synthesize speech."""

    def __init__(self, transport: Transport) -> None:
        self._t = transport

    def transcribe(
        self,
        audio: bytes,
        *,
        filename: str = "audio.webm",
        content_type: str = "audio/webm",
    ) -> TranscriptionResult:
        """Transcribe audio bytes to text.

        Args:
            audio: Raw audio bytes.
            filename: Filename hint for the server (determines codec detection).
            content_type: MIME type of the audio data.

        Returns:
            TranscriptionResult with text, language, and duration.
        """
        data = self._t.post_multipart(
            "/v1/voice/transcribe",
            files={"file": (filename, audio, content_type)},
        )
        return TranscriptionResult.model_validate(data)

    def synthesize(
        self,
        text: str,
        *,
        voice: VoiceId = "nova",
        speed: float = 1.0,
        response_format: AudioFormat = "opus",
    ) -> bytes:
        """Convert text to speech audio.

        Args:
            text: The text to synthesize (max 4096 chars).
            voice: TTS voice id.
            speed: Playback speed (0.25–4.0).
            response_format: Output audio format.

        Returns:
            Raw audio bytes in the requested format.
        """
        return self._t.post_binary(
            "/v1/voice/synthesize",
            json={
                "text": text,
                "voice": voice,
                "speed": speed,
                "response_format": response_format,
            },
        )

File: server/src/services/voice.ts
import OpenAI from "openai";
import { config } from "../config.js";

const openai = new OpenAI({ apiKey: config.OPENAI_API_KEY });

export type VoiceId =
  | "alloy"
  | "ash"
  | "ballad"
  | "coral"
  | "echo"
  | "fable"
  | "nova"
  | "onyx"
  | "sage"
  | "shimmer";

export type AudioFormat = "mp3" | "opus" | "aac" | "flac" | "wav" | "pcm";

export interface TranscribeResult {
  text: string;
  language: string | null;
  duration: number | null;
}

export interface SynthesizeOptions {
  voice?: VoiceId;
  speed?: number;
  response_format?: AudioFormat;
}

/**
 * Transcribe audio to text using OpenAI Whisper.
 */
export async function transcribe(
  audioBuffer: Buffer,
  filename: string = "audio.webm",
): Promise<TranscribeResult> {
  const file = new File([new Uint8Array(audioBuffer)], filename, {
    type: mimeFromFilename(filename),
  });

  const result = await openai.audio.transcriptions.create({
    model: "whisper-1",
    file,
    response_format: "verbose_json",
  });

  return {
    text: result.text,
    language: (result as any).language ?? null,
    duration: (result as any).duration ?? null,
  };
}

/**
 * Synthesize text to speech using OpenAI TTS.
 */
export async function synthesize(
  text: string,
  options: SynthesizeOptions = {},
): Promise<Buffer> {
  const voice = options.voice ?? "nova";
  const speed = options.speed ?? 1.0;
  const response_format = options.response_format ?? "opus";

  const response = await openai.audio.speech.create({
    model: "tts-1",
    voice,
    input: text,
    speed,
    response_format,
  });

  const arrayBuffer = await response.arrayBuffer();
  return Buffer.from(arrayBuffer);
}

function mimeFromFilename(filename: string): string {
  const ext = filename.split(".").pop()?.toLowerCase();
  const mimes: Record<string, string> = {
    webm: "audio/webm",
    ogg: "audio/ogg",
    mp3: "audio/mpeg",
    wav: "audio/wav",
    flac: "audio/flac",
    m4a: "audio/m4a",
    mp4: "audio/mp4",
  };
  return mimes[ext ?? ""] ?? "audio/webm";
}
